pintracker/optracker/operationtracker.go

```diff
@@ -0,0 +1,159 @@
+// Package optracker implements functionality to track the status of pin and
+// operations as needed by implementations of the pintracker component.
+// It particularly allows to obtain status information for a given Cid,
+// to skip re-tracking already ongoing operations, or to cancel ongoing
+// operations when contradictory ones arrive.
+package optracker
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	"github.com/ipfs/ipfs-cluster/api"
+
+	cid "github.com/ipfs/go-cid"
+	logging "github.com/ipfs/go-log"
+	peer "github.com/libp2p/go-libp2p-peer"
+)
+
+var logger = logging.Logger("optracker")
+
+// OperationTracker tracks and manages all inflight Operations.
+type OperationTracker struct {
+	ctx context.Context // parent context for all ops
+	pid peer.ID
+
+	mu         sync.RWMutex
+	operations map[string]*Operation
+}
+
+// NewOperationTracker creates a new OperationTracker.
+func NewOperationTracker(ctx context.Context, pid peer.ID) *OperationTracker {
+	return &OperationTracker{
+		ctx:        ctx,
+		pid:        pid,
+		operations: make(map[string]*Operation),
+	}
+}
+
+// TrackNewOperation creates and stores a new operation when no operation for
```

Comment:
No, existing operations are never returned because that would allow anyone to control an existing operation that they have not created and I want to explicitly disallow that. Once created, an operation is only in the hands of the creator (because it was returned to it) or by the tracker itself.

So here I disagree, nil means `nothing to do`. From the user point of view there are 2 cases only: 1) you get an operation and can do what you wanted 2) you don't get one and do nothing.

Reply:
Rewrote the docstring though